W3ChampionsStatisticService/Achievements/AchievementRepositoryHandler.cs

```diff
@@ -0,0 +1,272 @@
+            var seasonList = new List<int>();
+            foreach (Season s in seasons){seasonList.Add(s.Id);}
+            seasonList.Reverse();
+            return seasonList;
+        }
+
+        private List<Achievement> UpdateCurrentPlayerAchievementList(List<Achievement> currentAchievementsList){
+            if (currentAchievementsList == null) {currentAchievementsList = new List<Achievement>();}
+                var currentListIds = new List<long>();
+                foreach(Achievement achievement in currentAchievementsList){
+                    currentListIds.Add(achievement.Id);
+                }
+                foreach(long activeAchievementsId in ActiveAchievementIds){
+                    if(!currentListIds.Contains(activeAchievementsId)){
+                        switch (activeAchievementsId) {
+                            case 0:
+                            currentAchievementsList.Add(new MapWith25WinsAchievement());
+                            break;
+                            case 1:
+                            currentAchievementsList.Add(new Win10GamesWithATPartnerAchievement());
+                            break;
+                        }
+
+                    }
+            }
+            return currentAchievementsList;
+        }
+
+        private async Task<List<Matchup>> GetAllPlayerMatches(PlayerOverallStats playerOverallStats){
+            var playerMatches = new List<Matchup>();
+            var battleTag = playerOverallStats.BattleTag;
+            var playerRaceOnMapVersusRaceRatios = new List<PlayerRaceOnMapVersusRaceRatio>();
+            var seasons = ConvertSeasonsToSimpleList(playerOverallStats.ParticipatedInSeasons);
+
+            foreach(int s in seasons){
+                var playerRaceOnMapVersusRaceRatio = await _playerStatsRepository.LoadMapAndRaceStat(battleTag, s);
+                playerRaceOnMapVersusRaceRatios.Add(playerRaceOnMapVersusRaceRatio);
+                var seasonalMatches = await _matchRepository.LoadFor(battleTag, null, GateWay.Undefined, GameMode.Undefined, 100, 0, s);
+
+                foreach(Matchup matchup in seasonalMatches){
+                    playerMatches.Add(matchup);
+                }
+            }
+            return playerMatches;
+        }
+
+        private async Task<Achievement> UpdateCurrentPlayerAchievement(
+            Achievement playerAchievement,
+            PlayerOverallStats playerOverallStats,
+            string battleTag,
+            List<Matchup> matches,
+            bool isFirstRun
+            ){
+            if (matches == null){matches = await GetAllPlayerMatches(playerOverallStats);}
+            if (battleTag == null){battleTag = playerOverallStats.BattleTag;}
+            if (isFirstRun){playerAchievement.Counter = new Dictionary<string, int>();}
+        
+            var achievementProgressCounter = playerAchievement.Counter;
+            switch(playerAchievement.Id){
```

Comment:
i would put this logic in the achievement classes and make it an interface or abstract class. then you would only have to call a method like "IsAchieved(player, game)" or something like that (depending in what you need). Like that you could move the logic in the achievements and when we want to implement new achievements, we could just implement a class and be done with it. Issue also will be (right now): what happens if I add a new achievement? I would have to whipe and recalculate all achievements, right?

Reply:
You wouldn't need to recalculate all achievements. This will add the new one BUT we would need to grab all the player information again to run through it and look for the data we need.